MWT/how-to-container.rst

```diff
@@ -1,6 +1,66 @@
+
+The ``computeContentOptimalSize()`` method is called by the MWT framework in order to know the optimal size of the container.
+The optimal size of the container should be big enough so that each child can be allocated a size at least as big as its own optimal size.
+For example, the optimal width of a horizontal list is the sum of the optimal width of every child, and the optimal height is the maximum optimal height of all children.
+
+The container is responsible for computing the optimal size of every child. To do so, the ``computeChildOptimalSize()`` method should be called for every child.
+After this method is called, the optimal size of the child can be retrieved by calling ``getWidth()`` and ``getHeight()`` on the child widget.
+
+The ``Size`` parameter of the ``computeContentOptimalSize()`` method initially contains the size available for the container.
+An available width or height equal to ``Widget.NO_CONSTRAINT`` means that the optimal size should be computed without considering any restriction on the respective axis.
+Before the method returns, the size object should be set to the optimal size of the container.
+
+"blurry point": should the optimal size always be lower than the available size?
+
+Laying out the children of the container
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The ``layOutChildren()`` method is called by the MWT framework in order to lay out every child of the container.
+If a child is laid out outside the bounds of the container (partially or fully), only the part of the widget which is within the container bounds will be visible.
+
+The container is responsible for laying out each child. To do so, the ``layOutChild()`` method should be called for every child.
+Before this method is called, the optimal size of the child can be retrieved by calling ``getWidth()`` and ``getHeight()`` on the child widget.
+
+When laying out a child, its bounds have to be passed as parameter. The position will be interpreted as relative to the position of the container content.
+This means that the ``(0, 0)`` coordinates represent the top-left pixel of the container content and the ``(contentWidth-1, contentHeight-1)`` coordinates represent the bottom-right pixel of the container content.
+
+Providing children manipulation APIs
+------------------------------------
+
+The ``Container`` class introduces protected APIs in order to manipulate the list of children of the container.
+These methods may be overriden in the container subclass in order to make these APIs available for the user.
+
+Each of the following methods may be overriden individually:
+
+- ``addChild()``
+- ``removeChild()``
+- ``removeAllChildren()``
+- ``insertChild()``
+- ``replaceChild()``
+- ``changeChildIndex()``
+
+For example, the following snippet allows the user to call the ``addChild()`` method on the container:
+
+.. code-block:: Java
+
+	@Override
+	public void addChild(Widget child) {
+		super.addChild(child);
+	}
 
```

Comment:
Another notion that can be added (maybe later) is the possibility to change the default children visibility by overriding `setShownChildren()` and using `setShownChild()`.

Reply:
i forgot this method, i will add a few lines

Reply:
done